Add multi-threading to scheduler

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
The scheduler has been single threaded ever since its conception. While profiling the scheduler for performance, some bottlenecks seemed like they could be resolved via multi-threading. So, this is an attempt at the same.


#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
I've added a thread-pooling infrastructure to the scheduler using pthreads and parallelized 7 areas of the scheduler with it:
**query_nodes
query_jobs
dup_nodes
dup_resource_resv_array
check_node_array_eligibility
free_nodes
free_resource_resv_array**

Basic algorithm:
- The worker threads come up the start of a sched cycle and wait on a work queue.
- The main thread queues up tasks for the worker threads who pick it up and post results in a result queue

Before you ask, I didn't use function pointers for the thread-pool infrastructure because when I tested it, it turned out to be slower, I think the compiler cannot optimize the code as much with function pointers.

The next step could be to try and parallelize the eval_simple_selspec, the part where it goes through all nodes to see where the chunks fit. Although I didn't really see a bottleneck there during my testing, so we should first analyze if it'll be worth the effort.

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1344208899/New+pbs+sched+option+to+control+scheduler+multi-threading

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
My performance test submits around 10k jobs on a cluster of 10k nodes with calendaring turned on, backfill_depth = 20. Results (tested on a 4 core machine with hyperthreading):
Before multi-threading:
[test_mainline.log](https://github.com/PBSPro/pbspro/files/3473104/test_mainline.log)
Time taken by 100 sched cycles: 1903.9519248 (avg sched cycle time: ~19 seconds)

After multi-threading:
[test_mt.log](https://github.com/PBSPro/pbspro/files/3473106/test_mt.log)
Time taken by 100 sched cycles: 829.299150944 (avg sched cycle time: ~8.3 seconds)


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

agrawalravi90 commented:
@bhroam and @arungrover Appveyor finally passed, this PR is ready for review now, thanks.

agrawalravi90 commented:
@bhroam and @arungrover gentle reminder, thanks.

bhroam commented on src/scheduler/resource_resv.c:
```diff
@@ -304,27 +427,91 @@ dup_resource_resv_array(resource_resv **oresresv_arr,
+	th_task_info *task = NULL;
+	int num_tasks = 0;
+	int num_resresv;
+	int num_resresv_cpy;
```
Same as above, rename this to something with thread in it.

bhroam commented on src/scheduler/resource_resv.c:
```diff
@@ -304,27 +427,91 @@ dup_resource_resv_array(resource_resv **oresresv_arr,
 		log_err(errno, "dup_resource_resv_array",  "Error allocating memory");
 		return NULL;
 	}
+	nresresv_arr[0] = NULL;
```
Since you're mucking in this area of the code, change this to use \_\_func\_\_ and the macro

bhroam commented on src/scheduler/resv_info.c:
```diff
@@ -915,6 +916,10 @@ check_new_reservations(status *policy, int pbs_sd, resource_resv **resvs, server
+	err = new_schd_error();
+	if (err == NULL)
+		return -1;
+
```
Now that you allocate this and pass it to the dup functions, you need to free it someplace

bhroam commented on test/tests/performance/pbs_sched_perf.py:
```diff
@@ -249,3 +249,41 @@ def test_pset_fuzzy_perf(self):
+                a = {'Resource_List.select':
+                     str(nchunks) + ":ncpus=1:color=" +
+                     str(self.colors[i % 7])}
+                njobs = chunk_size / nchunks
```
colors is a list of strings, no need to str() it.

bhroam commented on test/tests/performance/pbs_sched_perf.py:
```diff
@@ -249,3 +249,41 @@ def test_pset_fuzzy_perf(self):
+        chunk_size = 100
+        total_jobs = 10000
+        while total_jobs > 0:
+            for i in range(chunk_size):
```
Why not do a range(10000)?

agrawalravi90 replied:
I wanted to submit a mix of large and small jobs and I also wanted to submit a small number of large jobs and a large number of smaller jobs. This is just how I was able to automate the number of chunks to be in the range of 1 to 100 (chunk_size) and still submit the jobs in the pattern that I wanted.

bhroam replied:
Oh, I didn't notice the jobs changing.  Yeah, different sizes of jobs is a good idea.

agrawalravi90 replied:
oh, good info, so I should perhaps also do a mix of users

bhroam commented on test/tests/performance/pbs_sched_perf.py:
```diff
@@ -249,3 +249,41 @@ def test_pset_fuzzy_perf(self):
+                    break
+
+        t1 = time.time()
+        # Run 100 sched cycles
```
Either explain why you're running 100 scheduling cycles or delete the comment.  The for loop is pretty self explanatory.

agrawalravi90 commented on src/scheduler/job_info.c:
```diff
@@ -493,187 +493,77 @@ static struct fc_translation_table fctt[] = {
+		return;
 	}
-	resresv_arr[num_prev_jobs] = NULL;
+	resresv_arr[0] = NULL;
```
right, so this code will be executed by the worker threads. They create their individual resresv_arr arrays, then inside query_jobs, the main thread creates the overall resresv_arr correctly by using num_prev_jobs. Refer to the code block below "/* Assemble job info objects from various threads into the resresv_arr */"

agrawalravi90 replied:
No, for query_jobs, the main thread has to assemble the individual resresv_arr arrays from different worker threads for exactly this reason. Please refer to the code block below "/* Assemble job info objects from various threads into the resresv_arr */"

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -671,6 +672,11 @@ scheduling_cycle(int sd, char *jobid)
 		"", "Starting Scheduling Cycle");
 
+	if (threads == NULL) {
+		if (init_multi_threading() != 0)
```
I had debated about this myself. I wanted the core scheduling library, or libpbssched to have the multithreading infrastructure, not the program pbs_sched. The scheduling library can be used by other programs for their own purposes, so I wanted it to be contained inside it. I didn't see any performance lag caused by the creation and deletion of threads every cycle, it takes microseconds, so I decided to keep it inside the library.

agrawalravi90 replied:
I'd thought about that as well, I didn't do it there because then I didn't know where I would gracefully kill the threads. Right now, it happens in end_cycle_tasks(), if we remove it from there, we'd have to put it in the pbs_sched program's main, and I wanted to avoid that.

So I'd rather we keep it the way it is since it doesn't hurt performance, unless you can think of a way to keep them inside libpbssched and start and end them cleanly.

arungrover replied:
Well, if you add the creation of threads in schedinit() (or something similar) you can also add a schedexit() in libpbssched which can be called by pbs_sched when it exits.
I don't see how it is helping us by recreating threads every cycle. IMO, Thread pool of worker threads should persist throughout the life of the process. Also, putting code in end_cycle_tasks does not guarantee any graceful cleanup of threads anyways, so I do not think it would be any worse.

agrawalravi90 replied:
hmm .. ok, i looked at how the pbs start/stop script works, and it sends TERM to pbs_sched, so ya, putting it in end_cycle_tasks() won't really help terminate it gracefully. I'll do what you suggested and create the threads in schedinit(), and add graceful termination inside SIGTERM handler instead.

bhroam replied:
Actually what you had before is fine.  During a scheduling cycle, the scheduler blocks all signals.  If you keep what you have, you need to terminate the threads in the QUIT part of schedule as well.

agrawalravi90 commented on src/scheduler/job_info.c:
```diff
@@ -493,187 +493,77 @@ static struct fc_translation_table fctt[] = {
+
+	/* Move to the linked list item corresponding to the 'start' index */
+	for (cur_job = jobs, i = 0; i < sidx && cur_job != NULL; cur_job = cur_job->next, i++)
+		;
```
I could but is that any better than passing in sidx? passing in pointer would mean that the main thread would have to serially traverse the list to find the pointer to pass to the worker threads. Passing in sidx lets the worker threads find their start pointer in parallel.

agrawalravi90 commented on src/scheduler/job_info.c:
```diff
@@ -949,15 +835,244 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+	}
+	jinfo_arrs_tasks = malloc(num_tasks * sizeof(resource_resv **));
+	if (jinfo_arrs_tasks == NULL) {
+		log_err(errno, "query_jobs", "Error allocating memory");
```
Sorry, but which one?

agrawalravi90 commented on src/scheduler/multi_threading.c:
```diff
@@ -0,0 +1,225 @@
+		result_queue = new_schd_queue();
+		if (work_queue == NULL || result_queue == NULL) {
+			snprintf(log_buffer, sizeof(log_buffer), "Error malloc'ing thread memory");
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
```
I kind of do, I'm logging once right after the 2 consecutive malloc calls, I didn't want to check for NULL one at a time. But let me know if you'd prefer that I check them individually.

bhroam replied:
It's the way the scheduler does malloc checking.  They usually happen in the constructor right after the malloc() call.  I don't see any problem checking them individually.

agrawalravi90 commented on src/scheduler/multi_threading.c:
```diff
@@ -0,0 +1,225 @@
+		num_cores = sysinfo.dwNumberOfProcessors;
+#else
+		num_cores = sysconf(_SC_NPROCESSORS_ONLN);
+#endif
```
Hmm .. I see your point. A few things:
- the threads sleep until work is available for them, they aren't spin looping, so cores won't be consumed until needed.
- the work is segregated in bug chunks for the worker threads. So, if a site has less than 1024 nodes, only 1 worker thread will be woken up for the node functions. Similarly, if there are less than 1024 jobs in the queue at a time, only 1 worker thread will be woken up for the job functions. So, unless a site has a large number of jobs/nodes, not all worker threads will be used.
- In my testing with a million jobs and 50k nodes on a machine with 8vcpus (4 cores, 2 hyperthreads each), I still didn't see the overall CPU% go beyond 400%.

So, we can make this a configurable, but I don't like the idea of using just half the cores by default. Maybe we should use N - 1 cores? leave one for the server since it's single threaded? @billnitzberg seeking advice from you as well, thanks!

bhroam replied:
The problem with using N-1 is that we're planning on horizontally scaling multiple PBS components (e.g. multi-server).  We don't want to hog all the CPUs for the scheduler, or worse, have the PBS components fight for CPU cycles and slow things down.  We usually put multiple schedulers on one machine (they don't have to), so using all of the CPUs will clash with different schedulers.  I definitely think we should make it configurable.  I'm open to ideas about the default though.  I am curious what @billnitzberg and @subhasisb has to say about this.

billnitzberg replied:
A couple ideas:
(1) Make it configurable & set the default = 1. This allows rolling out this (risky) change slowly, only to those who explicitly set a value > 1. We can change the default in a future release.
(2) Benchmark a handful of representative workloads on a machine with lots (>20) cores, varying the numbers of threads, and see where the knee in the curve is and use that as the default. I suspect somewhere between 2 - 20 threads will be a good default.
(3) What does the server do for TPP threads? What about Postgresql? Both these will likely run on the same node as the Scheduler.

arungrover replied:
I like your first approach although I think the default should not be 1. Making it 1 will make it slower than what it is today. Maybe, by default, we don't spawn any threads at all and keep the behavior as it is today.

agrawalravi90 commented on src/scheduler/node_info.c:
```diff
@@ -1213,20 +1457,67 @@ dup_nodes(node_info **onodes, server_info *nsinfo,
+				free(tdata);
+				free(task);
+				i++;
+			}
```
There isn't really much that the worker threads leave for the main thread to do in this case. Worker threads directly modify the new node info array for their respective indices and dup data into it. The only thing of use for the main thread is the error bit. It still has to loop over the result queue, wait until all worker threads post their results, and only then move on.

bhroam replied:
Interesting.  I guess this just didn't match up with how I imagined how threads worked.  I thought you had some master process that spun up threads and then collected results and processed them.  I guess this works just as well.  The threads themselves know what portion of the problem they are working on, and place the results back in the right place.  One downside to this is if threads ever get confused on portion of the problem they are working on and overlap with another thread, it will be really hard to debug.  Debugging threads is going to be difficult in any case.  Just curious, what made you go with this model instead of the master/client model?

agrawalravi90 replied:
I just went for the most efficient way. If we went with the master thread assembling the different dup'd fractions, it'd be worse w.r.t. performance and memory as the worker threads would send their results in their own dup arrays, and the master thread would iterate over their individual results and collate them. If the worker threads got confused, we'd see a mess in the final array regardless of which approach we go with. It's a very simple partitioning logic though, so they really shouldn't get confused.

bhroam replied:
If you go with Arun's suggestion to make the thread pool code into a generic library, would that change your mind?  I don't know enough about threads to know one way or the other.

agrawalravi90 replied:
Depending on the task, the worker threads might do very different things. So, even if we made it generic, depending on the task at hand, the master and worker threads will still have different levels of responsibilities, the master thread may do more post-processing for some tasks and less for others.

bhroam replied:
Yeah we'd have to use callbacks to reassemble things as well.

agrawalravi90 commented on src/scheduler/node_info.c:
```diff
@@ -5831,12 +6122,80 @@ is_exclhost(place *placespec, enum vnode_sharing sharing)
+							|| sim_exclhost(resresv->server->calendar, resresv, node) == 0) {
+						int j;
+
+						for (j = 0; node->hostset->ninfo_arr[j] != NULL; j++) {
```
good catch! I didn't realize that scr.ineligible is being read as well. I think there are a few choices:
1) Don't change anything. I might be wrong about this but from what I understand, even if one thread marks a node src.ineligible while the other is already inside the if ((!node->nscr.ineligible))  block, the other thread will just do some extra work before marking the nscr.ineligible bit. If this is True, then, again I might be wrong about this, but the fact that we send work out in big chunks should help nodes of the same hostset get handled by a single thread. So, this shouldn't happen for many nodes. But please do correct me if my understanding is wrong.
2) Remove the 'if (node->hostset != NULL)' block. This will probably cause some performance degradation, but hopefully not too much as the threads will just have to check each and every node instead of marking all nodes of same hostset as ineligible. Again, I'm not entirely sure if this block can be removed without affecting anything functionally, so please do clarify.
3) If neither of the above can be done, then I guess we'll just have to remove multithreading for this function, as it seems to have iterative dependencies.

Please advise.

bhroam replied:
So you are correct, the block of code will do extra work and mark the scr.ineligible bit.  What I don't know is if there can be a problem between the instructions of the if statement.  The worst case scenario will be all N nodes in the host set will be in N different threads.  We might have N different threads doing the same work and marking the same nodes as ineligible at the same time.  I don't see this as a degradation of performance since before it was iterative as long as you don't see it being a problem.  #2 can't be done.  We used to not have that check, and performance was a dog for systems without hostests (like an ICE).  Things sped up significantly after we added that check.  You likely want to add a comment saying that the different nodes in the hostset might be in different threads.  Someone in the future might come along and do something that is harmful.

agrawalravi90 replied:
sorry for the delay in getting back to this @bhroam . So, should I just add a comment there saying that different nodes in hostset could be processed by different threads, in which case there might be some duplication of work?

bhroam replied:
Yeah, that sounds about right.  The main idea I want to portray is that the different nodes in the hostsets might be handled by different threads.  If someone wants to modify other nodes in the hostsets other than marking the ineligible bit, they should be careful.

agrawalravi90 commented on src/scheduler/queue.c:
```diff
@@ -0,0 +1,178 @@
+ * @retval NULL for malloc error
+ */
+schd_queue *
+new_schd_queue(void)
```
hmm .. ok, I'm renaming it to 'ds_queue'. Please let me know if you can think of a better name for it.

arungrover commented on src/scheduler/queue.c:
```diff
@@ -0,0 +1,177 @@
+ * @retval NULL for malloc error
+ */
+ds_queue *
+new_ds_queue(void)
```
Can I recommend a change from ds_queue to thread_comm_queue or tc_queue. I'm not sure it needs 'ds' there in the name.

agrawalravi90 replied:
So, this is a generic queue implementation, parts of the code other than threading can also make use of it. I added 'ds' in there because Bhroam was concerned that people might confuse this with PBS queues.

arungrover replied:
In that case, how about renaming it to work_queue or pbs_queue?

bhroam replied:
I think we should name it something completely separate.  pbs_queue is definitely wrong because pbs queues are what jobs are submitted into.  work_queue is the name of our default queue, so that might also be confusing.  BTW, Ravi, what does ds stand for?

agrawalravi90 replied:
ds = data structure

I agree with bhroam that work_queue and pbs_queue can both be confusing.

you guys don't like ds_queue?

agrawalravi90 replied:
@bhroam not really, I copied this data structure code from another project where I was using it as a normal queue, it's really just a generic queue.

arungrover commented on src/scheduler/queue.c:
```diff
@@ -0,0 +1,177 @@
+ * @return void
+ */
+void
+free_ds_queue(ds_queue *obj)
```
can you rename obj to queue because everywhere else you are using obj as the object to store

arungrover commented on src/scheduler/queue.c:
```diff
@@ -0,0 +1,177 @@
+	if (queue->front == queue->rear) /* queue is empty */
+		return NULL;
+
+	return queue->content_arr[queue->front++];
```
Does this mean that objects will be there in the queue even after they are dequeued? I think we should memcpy and realloc after dequeuing some number of objects.

agrawalravi90 replied:
Thanks for catching this, right now the queue can grow forever! But I think it might be better  if we reset the front and rear pointers when the queue gets empty. This will allow better use of it once the queue reaches a big enough size and will prevent the queue from growing forever. So, it'll get used like a buffer. What do you think?

arungrover replied:
I like this Idea

bhroam commented on test/tests/performance/pbs_sched_perf.py:
```diff
@@ -271,8 +271,7 @@ def test_many_jobs_with_calendaring(self):
             for i in range(chunk_size):
                 nchunks = i + 1
```
Instead of this, you could do for i in range (1, chunk_size + 1) and then you don't need the nchunks variable

bhroam commented on src/scheduler/data_types.h:
```diff
@@ -194,21 +194,21 @@ struct th_data_nd_eligible
 
 struct th_data_dup_nd_info
 {
+	unsigned int err:1;
```
I'd call this error.  Currently the only time we have a variable called 'err' it is a schd_error.  You even call a member err later that is of type schd_error.

agrawalravi90 commented on src/scheduler/constant.h:
```diff
@@ -168,6 +168,18 @@ enum resval_cmpflag
 	CMP_CASELESS
 };
+
+enum thread_task_type
```
To be honest, I don't think any other PBS daemon will benefit from multi-threading like this. PBS Server doesn't do heavyweight calculations like the scheduler does, most of its cpu time is spent sending and receiving requests. The mom daemon just launches jobs and monitors them, not much to process there either. pbs_comm has its own thing already.

So, given how using callbacks degrades performance, we should only use them if they'll benefit us in a significant way, and I just don't see that ROI. Please let me know what you think.

arungrover replied:
You have a fine implementation of thread pool in place. It would be a shame if other components can not make use of it without making significant changes to it.
I understand server may not have as many places it can use a thread pool as scheduler does, but, there could be cases where it might need it in future (like sending mom hooks, clearing up job-history, maybe implement a notification service or do other things).
One of the reasons deleting jobs is slow is because mom is trying to scan through the proc file system for every session ID it knows (serially), with thread-pool it can use threads to do that. Mom could also use this to monitor jobs as well.

IIRC, you mentioned that scheduler takes around half the time to process a million jobs when this thread implementation is being used (roughly from 6mins to 3 mins), but when you use function pointers it drops by 40 seconds. While 40 seconds is a lot of degradation for a 3 min cycle, but I think we have still improved it by more than 2 mins out of 6 mins which is significant.

agrawalravi90 replied:
hmm .. so we don't really need to change the algorithm to use callbacks in order to make it generic, right? we can just add more tasks to the task list and call their corresponding task handlers, which could be non-scheduler. So, if we ever have a need for it, we can just put the code in a common library instead of the scheduler. Does that work?

arungrover replied:
I don't think just adding the tasks to the list and calling their respective methods will do it. a non-scheduler component will then be compiling with references present in sched library.

agrawalravi90 replied:
ok ya, you are correct .. I don't know man, I really don't want to make a change which will hurt performance when the whole purpose of this PR is to improve performance. If some other part of PBSPro needs multi-threading, it'll be better for performance to just create the infrastructure for that part separately, it's really not that much code, most of the code change in this PR is the task handlers themselves. We should also keep in mind that, depending on the use case, the other parts of PBS might need a different multi-threading approach than the one used here with the worker/result queues. So, I'd like to keep this as it as for now. If you still think we should go the callback route, then we can ask others for opinions.

arungrover replied:
Ok, do you have the function pointer implementation of yours that you can share. It will be interesting to see why is function pointer slower than managing a switch statement.
I agree that since our discussion isn't moving anywhere, It would be better to seek opinions from others.

subhasisb replied:
I agree with Arun here, in general (without actually seeing the specific implementation or considerations here). It would be great if you can make the thread pool code generic (and into a library) that can then be utilized by anybody (daemons or not) wanting to use a thread pool. I don't know what that would be the impact on the timeline for this PR, but we should consider it. If we cannot make that refactoring within this PR's timeframe, we could file another refactoring RFE to do it shortly after?

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -949,15 +837,248 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+	for (i = 0, jidx = num_prev_jobs; i < num_tasks; i++) {
+		if (jinfo_arrs_tasks[i] != NULL) {
+			for (j = 0; (resresv = jinfo_arrs_tasks[i][j]) != NULL; j++) {
+				resresv_arr[jidx++] = resresv;
```
Why bother with the resresv variable?  You only use it once and setting it in the while portion of the for loop is strange

agrawalravi90 replied:
I use it twice right? once to check that resresv != NULL, and once to add it to the resresv_arr.

bhroam replied:
You don't really use it twice.  When you check resresv != NULL, you're also setting resresv.  You are using job_arr_tasks[i][j] there as well.  I'd just use job_arrs_tasks[i][j] twice.

subhasisb commented on src/scheduler/node_info.c:
```diff
@@ -240,52 +303,104 @@ query_nodes(int pbs_sd, server_info *sinfo)
+		while (ds_queue_is_empty(result_queue))
+			pthread_cond_wait(&result_cond, &result_lock);
+		pthread_mutex_unlock(&result_lock);
+		while (!ds_queue_is_empty(result_queue)) {
```
Yes the unlock seems placed wrongly. The cond wait will re-aquire result_lock atomically before returning, and we should hold onto the lock while working on the shared ds.

agrawalravi90 replied:
Thanks for catching this guys, I'm surprised I didn't see any issues because of this during my testing!

subhasisb commented on src/scheduler/node_info.c:
```diff
@@ -160,6 +160,63 @@
+
+		if (node_in_partition(ninfo, sinfo->partitions)) {
+			/* get node info from mom */
+			if (talk_with_mom(ninfo)) {
```
talk_with_mom() might not be thread safe. Especially, it calls open_rm() which does not look like thread safe. Either we ensure that it is thread safe all through the call stack or we apply locks around it. However, serializing around an IO operation is generally a bad idea and could kill any performance benefit from the parallelization itself.

agrawalravi90 replied:
thanks for catching this!

subhasisb commented on src/scheduler/node_info.c:
```diff
@@ -637,13 +776,80 @@ void
+		pthread_mutex_lock(&result_lock);
+		while (ds_queue_is_empty(result_queue))
+			pthread_cond_wait(&result_cond, &result_lock);
+		pthread_mutex_unlock(&result_lock);
```
the unlock does not sound correct here. cond_wait will re-acquire locks atomically at return. You are not using that before checking ds_queue_is_empty in the first while loop - just doing it after the lock...besides the lower while loop seems unprotected in its use of result_queue

agrawalravi90 replied:
Thanks for catching this Subhasis!

subhasisb commented on src/scheduler/node_info.c:
```diff
@@ -160,6 +160,63 @@
+	num_nodes_chunk = end - start + 1;
+
+	if ((ninfo_arr = (node_info **) malloc((num_nodes_chunk + 1) * sizeof(node_info *))) == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
```
It will be great if you can clearly demarcate which of the functions are meant to be called from the worker threads and which are the main thread only. That way future developers can keep them safe. It might be beneficial to move all "threaded" functions to one area of the code file (say the end of the file) - with a clear demarcation in between the threaded and non-threaded code. (even better if you can move it to a separate file)

If this function is threaded, then the log_err might not be safe. We must used a thread local buffer for the log message

subhasisb commented:
@agrawalravi90 can you kindly run hellgrind on your code and run all your test cases (multiple times) over it. Hellgrind will attempt to point out synchronization errors and race conditions.

agrawalravi90 commented:
@bhroam and @arungrover I am still in the process of addressing your comments. In the meantime, I've added a new sched attribute to control the number of threads, added link to EDD in the description, request you to review that commit, I will keep chomping on your comments. Thanks!

arungrover commented on src/lib/Libecl/ecl_verify_values.c:
```diff
@@ -1949,3 +1949,37 @@ verify_value_preempt_sort(int batch_request, int parent_object, int cmd,
+		return PBSE_BADATVAL;
+
+	numval = strtol(pattr->value, &endp, 10);
+	if (*endp != '\0' || numval <= 0 || numval >= 100000)
```
Isn't 100000 too big a number for configuring the number of worker threads? I'd suggest we do not check for max value, rather let scheduler set it to as the number of cores present in the system when the thread count given is more than that.

arungrover replied:
If you chose to make this change then this verify function is not needed, you can use verify_value_non_zero_positive instead

arungrover commented on src/scheduler/data_types.h:
```diff
@@ -194,7 +194,7 @@ struct th_data_nd_eligible
 struct th_data_dup_nd_info
 {
-	unsigned int err:1;
+	unsigned int error:1;
```
Do you still want to keep it as a bit field? It will loose the flexibility of classifying errors.

arungrover commented on src/scheduler/pbs_sched.c:
```diff
@@ -159,6 +160,11 @@ extern char *msg_startup1;
+
+	thid = (int *) pthread_getspecific(th_id_key);
+	if (thid != NULL && *thid != 0)
+		pthread_exit(NULL);
```
Does this mean that a specific thread will exit? If so, then upon receiving SEGV more than 5 min apart we fork() ourselves and let the child continue. 
How will threads behave when scheduler forks and wouldn't pthread_exit() reduce a worker thread out of the process and scheduler will not be aware of this.

agrawalravi90 replied:
this will make all worker threads exit, it's a bad idea to fork() with threads, it's safer to create them again, which should happen when the new process executes afresh. But I noticed that I'm not calling kill_threads() here, which should get called before the fork() as well.

arungrover commented on src/server/req_manager.c:
```diff
@@ -1846,6 +1845,36 @@ mgr_sched_unset(struct batch_request *preq)
+			GetSystemInfo(&sysinfo);
+			num_cores = sysinfo.dwNumberOfProcessors;
+#else
+			num_cores = sysconf(_SC_NPROCESSORS_ONLN);
```
It is probably going to make it too complicated but what happens where there are multiple-schedulers involved? would it set it to the same value for all scheds?
If you don't want to change it, then at least move this assignment to scheduler because scheduler might end up running on a machine where the server isn't present, so the cores won't match.

agrawalravi90 commented:
@arungrover  and @bhroam sorry guys, but I got feedback on the EDD for the new sched attr, and I think I'm going to change it to a pbs_sched command line option instead of a qmgr attr. I'll not change the code right now, will wait to get consensus on the design first, should have done that in the first place! :) Anyways, please don't review the latest commit yet as i might revert the sched attr changes. Sorry for the mess!

agrawalravi90 commented on src/scheduler/globals.c:
```diff
@@ -147,6 +148,19 @@ int got_sigpipe;
 int	second_connection;
 
+/* Stuff needed for multi-threading */
+pthread_mutex_t work_lock;
```
@bhroam what do you think?

bhroam replied:
You can't put global variables in a header file, only extern references to them.  We'd have to move them into a .c file.  If we think we might move this code out of the scheduler in the future, then we should treat is more like a library and move these variables elsewhere, otherwise I think they are fine where they are.  This is a file meant to hold globals.

agrawalravi90 commented on src/scheduler/data_types.h:
```diff
@@ -165,6 +174,83 @@ typedef sch_resource_t usage_t;
+
+struct th_data_dup_nd_info
+{
+	unsigned int err:1;
```
It should be pretty easy to change if we have a need for this in the future, how about we change it if/when it's needed?

arungrover replied:
why not now? all we have to do is that we make it int and assign it a generic error number (which is most cases is going to be a malloc error)

agrawalravi90 replied:
thinking about this more, i think if there's a need to convey more information than just "error happened", then we should be passing a schd_error object, which I actually do use for check_node_array_eligibility, its thread data type 'th_data_nd_eligible' has a schd_error object.

arungrover replied:
Great! in that case, can you make it consistent across other datastructures?

agrawalravi90 replied:
I think I'm not explaining this well. Each 'th_data_' object is just a way to pass data to a work task. Since each work task is unique, each data object is also unique, this is very much like different functions requiring specific arguments. So, what you are asking me to do is similar to saying that pass an argument to a function which it won't use today but might use in the future. We don't pass a schd_error argument to all scheduler functions, right? Similarly, not all functions return multiple return values right? Most of them just return a 0 or a 1, which is exactly what I want to achieve with the 'error' bit. Do you see what I'm saying now?

arungrover replied:
Ok I understand your point. What I do not understand is that what benefit do we get if it is kept as a bit field in some structures.

agrawalravi90 replied:
As I said, there was a need to convey whether or not an error happened for some tasks, just like some functions return 0 or 1. A bit field serves that purpose, that's why I used it. Why spend more memory if we don't need to?

arungrover replied:
I don’t think you are saving any memory by making it a bit field. Your structure will be padded. Any other reason?

agrawalravi90 replied:
No other reason. There are many other scheduler structs which use this syntax for boolean flags, can you please clarify why you are being so particular about not using it here?

arungrover replied:
You use bit fields when you want to more than one thing at a time. Many other data structures in scheduler use it to convey flags (like user limit, group limit etc.) they don’t use it to convey errors. If errors were kept as flags we will be modifying structure every time there is a new type of error added. It just does not make any sense.

agrawalravi90 replied:
but it's not really supposed to convey the type of error that happened, it's just like a return value. I'll do 2 things, rename it to "retval" and convert it to an int, does that sound good? Please try to look at this struct as that function's arguments and return type, it will not be used anywhere else except that one function.

agrawalravi90 commented on src/scheduler/job_info.c:
```diff
@@ -949,15 +837,248 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+	}
+	if (th_err) {
+		pbs_statfree(jobs);
+		free_resource_resv_array(resresv_arr);
```
so we don't break out of that loop, we get results from all worker threads, free tdata and task, and then reach this if block. So, there shouldn't be a leak.

agrawalravi90 commented on src/scheduler/multi_threading.c:
```diff
@@ -0,0 +1,233 @@
+		threads = malloc(num_threads * sizeof(pthread_t));
+		if (threads == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			return 1;
```
actually these will get free'd up in schedexit(), which should get called when scheduler bails from here. But it's safer to free them up so I'll do it, thanks.

agrawalravi90 commented on src/scheduler/multi_threading.c:
```diff
@@ -0,0 +1,233 @@
+		}
+	}
+
+	ts = get_avl_tls();
```
So, after i turned off single threaded mode, i started seeing this leak. I don't remember now exactly how this is used, maybe @subhasisb can clarify?

agrawalravi90 commented on src/scheduler/node_info.c:
```diff
@@ -160,6 +160,63 @@
+				/* failed to get information from node, mark it not free for this cycle */
+				ninfo->is_free = 0;
+				ninfo->is_offline = 1;
+				schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_NODE, LOG_INFO, ninfo->name,
```
great catch, thanks!

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -671,6 +672,11 @@ scheduling_cycle(int sd, char *jobid)
 
+	if (threads == NULL) {
+		if (init_multi_threading() != 1)
+			return 1;
```
this was a mistake, thanks for catching it

agrawalravi90 commented on src/scheduler/node_info.c:
```diff
@@ -240,52 +303,104 @@ query_nodes(int pbs_sd, server_info *sinfo)
+	j = 0;
+	chunk_size = num_nodes / num_threads;
+	chunk_size = (chunk_size > 1024) ? chunk_size : 1024;
+	while (num_nodes > 0) {
```
I really wanted to do that as well, but it won't really reduce much code, and it will introduce many if else conditions as each task's data structure is different, has different members etc., so I think that one function will end up being a mess.

arungrover replied:
Right, how about we have one common function to place work and notify threads and then one function to dequeue results back. Currently, there is a lot of repetition in many places.

agrawalravi90 replied:
I'm talking about that function only Arun. Just to give you an idea, that one function's argument list will look something like this:

void send_work(int ct_left, enum thread_task_type task_type, node_info **onodes, node_info **nnodes, server_info *sinfo, unsigned int flags, struct batch_status *nodes, resource_resv **oresresv_arr, resource_resv **nresresv_arr, queue_info *nqinfo ..... (so on)

and this will need to be expanded as we add more tasks. The function body will similarly be quite a mess.

I can put the following 4 lines in a function, so I'll do that, please let me know if you can think of a clean way to generalize anything else.
pthread_mutex_lock(&work_lock);
ds_enqueue(work_queue, (void *) task);
pthread_cond_signal(&work_cond);
pthread_mutex_unlock(&work_lock);

agrawalravi90 commented:
ok, i've reverted the sched attr changes and made it a pbs_sched option, I think I've also addressed all of your review comments, I'll go through them once again, but this is ready to be reviewed again. thanks!

agrawalravi90 commented:
Thanks for your inputs Subhasis. Making the implementation generic shouldn't take a lot of time actually. But did you read the comment where I mentioned that making it a generic, function pointer based implementation degrades the performance by a non-trivial amount? Otherwise of course it's the right thing to do and I can do it without much effort, I just want us to think about whether it's worth having a generic library which performs slower than specific implementations, afterall the main goal here is performance.


Get Outlook for Android<https://aka.ms/ghei36>
________________________________
From: Subhasis Bhattacharya <notifications@github.com>
Sent: Tuesday, August 27, 2019 9:43:22 AM
To: PBSPro/pbspro <pbspro@noreply.github.com>
Cc: Ravi Agrawal <ravi.agrawal@altair.com>; Mention <mention@noreply.github.com>
Subject: Re: [PBSPro/pbspro] Add multi-threading to scheduler (#1228)


@subhasisb commented on this pull request.

________________________________

In src/scheduler/constant.h<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FPBSPro%2Fpbspro%2Fpull%2F1228%23discussion_r317886581&data=02%7C01%7Cravi.agrawal%40altair.com%7Cea3594f35d754bd9e96308d72aa4e816%7C2bae5b570eb848fbba47990259da89d2%7C0%7C0%7C637024760069031119&sdata=gTsdlhNFz2%2FwH3SbYJMZsXaRYAE1bdB%2FF7wRmdVu5bU%3D&reserved=0>:

> @@ -168,6 +168,18 @@ enum resval_cmpflag
        CMP_CASE,
        CMP_CASELESS
 };
+
+enum thread_task_type


I agree with Arun here, in general (without actually seeing the specific implementation or considerations here). It would be great if you can make the thread pool code generic (and into a library) that can then be utilized by anybody (daemons or not) wanting to use a thread pool. I don't know what that would be the impact on the timeline for this PR, but we should consider it. If we cannot make that refactoring within this PR's timeframe, we could file another refactoring RFE to do it shortly after?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FPBSPro%2Fpbspro%2Fpull%2F1228%3Femail_source%3Dnotifications%26email_token%3DABC45G2UKS7M6LYEZ5OTFX3QGSSWFA5CNFSM4IJX7AH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCX2LDI%23discussion_r317886581&data=02%7C01%7Cravi.agrawal%40altair.com%7Cea3594f35d754bd9e96308d72aa4e816%7C2bae5b570eb848fbba47990259da89d2%7C0%7C0%7C637024760069041116&sdata=qxLwja4SEmd2r4kebu6AC6FA3LH%2B8muRq1j%2FiAhWZH4%3D&reserved=0>, or mute the thread<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABC45G6NWVRR7NOWJ6WEDTLQGSSWFANCNFSM4IJX7AHQ&data=02%7C01%7Cravi.agrawal%40altair.com%7Cea3594f35d754bd9e96308d72aa4e816%7C2bae5b570eb848fbba47990259da89d2%7C0%7C0%7C637024760069041116&sdata=0v1sfuEARNLHmIEge3UeUugCKHoE6a8SS%2BPjfDrRFcQ%3D&reserved=0>.

subhasisb commented:
> Thanks for your inputs Subhasis. Making the implementation generic shouldn't take a lot of time actually. But did you read the comment where I mentioned that making it a generic, function pointer based implementation degrades the performance by a non-trivial amount? Otherwise of course it's the right thing to do and I can do it without much effort, I just want us to think about whether it's worth having a generic library which performs slower than specific implementations, afterall the main goal here is performance.

Hey Ravi, i did see that comment u made earlier. However, i do not understand why a function pointer (or callback type) implementation would make any difference in performance. Perhaps, we can look at that specific implementation and see why you are seeing any difference?

agrawalravi90 commented:
> Hey Ravi, i did see that comment u made earlier. However, i do not understand why a function pointer (or callback type) implementation would make any difference in performance. Perhaps, we can look at that specific implementation and see why you are seeing any difference?

I don't have that code any longer, I didn't realize that I'll need it. I had done performance profiling of the function pointer code and there wasn't anything special happening there in terms of the logic, no new bottlenecks. I believe that the performance degradation happens because the compiler cannot optimize the code as well when function pointers are involved. When I had compiled this code without any compiler optimizations and test it for that workload (1M jobs, 50k nodes), 1 sched cycles took 1.5-2 minutes longer, so that convinced me that compiler optimizations can be the cause of the degradation seen with func pointers code.

Anyways, is it okay if we file a separate task to explore a generic function pointer implementation of this, test its performance etc. ?

agrawalravi90 commented:
Actually, this is probably the best time to decide if the function pointer approach is worth it, so ill make the change in a separate branch and test it out and let you guys know.

subhasisb commented:
> Actually, this is probably the best time to decide if the function pointer approach is worth it, so ill make the change in a separate branch and test it out and let you guys know.

And, thinking further, if function pointers are a problem with optimization, it probably hurts a whole bunch of our code base, not just this, and we should have a "solution" to that optimization problem.  However, in general, we should tilt a tad bit towards modularity (code reuse) etc. in the balance between modularity and performance. (though your numbers do sound very significant)

agrawalravi90 commented:
@arungrover and @subhasisb , I've implemented the function pointer approach in a different branch, here's the commit: https://github.com/agrawalravi90/pbspro/commit/b25b8a00aa8df89402a562348794b1fdb796f8eb

I'm still testing it out to compare performance with the other approach, but please do review this and let me know if you see anything there that might explain the performance degradation. Rest, I'll do profiling and try to find any bottlenecks. Thanks.

agrawalravi90 commented:
@arungrover and @subhasisb , I've been running the performance test that I wrote for this PR and been consistently seeing difference of around a 100 seconds between function pointer approach and without (~680 seconds vs ~780 seconds). I can't really explain it, I don't see any new bottlenecks in the function pointer approach, but I'll profile more and see if I can find anything. In the meantime, please let me know your thoughts.

subhasisb commented:
Thanks for the test @agrawalravi90 - i found this nice discussion on the net: https://stackoverflow.com/questions/2438539/does-function-pointer-make-the-program-slow

Looks like function pointers can indeed slow down some compiler optimizations in some cases but i am seeing that a overall consensus on that page is about the flexibility gains largely outweigh the minor setbacks. 

So, back to your test results, the setback seems to be a 7-8% degradation over the direct call approach, right? If so, perhaps it is negligible compared to the gain in flexibility and modularity. 

On a more academic level, is there a faster way to implement callbacks in C other than function pointers - PBS code uses function points for everything, starting from the lifeline work-task to IFL indirections to network IO routine switching, virtually everything. So, just eliminating one may not make any big difference?

agrawalravi90 commented:
@bhroam and @arungrover I've made some fixes to the MT functions to make sure that they work in a serial fashion when num_threads is 1, and added the pbs.conf option PBS_SCHED_THREADS to control the thread count. Please review the latest commit(s). Thanks!

agrawalravi90 commented:
Thanks for catching that Mike, I've addressed your comment, please let me know if it looks okay now.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -867,7 +867,7 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
 	int    rc = 0;
 	FILE  *savlog;
-	static char slogbuf[LOG_BUF_SIZE];
+	char slogbuf[LOG_BUF_SIZE];
```
I believe there is a performance advantage to keeping this as a static so that it doesn't have to be reallocated on the stack each time log_record is called.

agrawalravi90 replied:
I had given some thought to that and researched online, and according to what I found, it seemed like the performance hit would be negligible. Here's a thread which talks about this:
https://stackoverflow.com/questions/3730000/can-static-local-variables-cut-down-on-memory-allocation-time

So, I'd like to keep this change, otherwise we'd have to create a thread-safe version of log_record() and have the scheduler use that, which doesn't seem worth it. But please let me know if you'd like me to spend more time on this.

mike0042 replied:
No need to spend more time on this. Thanks for the analysis and response, I'm OK with leaving this as you have it.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -996,84 +997,109 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
-		pthread_cond_signal(&work_cond);
-		pthread_mutex_unlock(&work_lock);
+		if (tdata->error || tdata->oarr == NULL)
+			return NULL;
```
You need to call pbs_statfree() (and other places)

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -996,84 +997,109 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
-			tdata = (th_data_query_jinfo *) task->thread_data;
-			if (tdata->error)
+		for (j = 0, jidx = num_prev_jobs; tdata->oarr[j] != NULL; j++) {
+			resresv_arr[jidx++] = tdata->oarr[j];
```
Does tdata->oarr need to be freed?

bhroam replied:
I'm not talking about the jobs themselves, I'm talking about the tdata->oarr array itself (the array of pointers).

agrawalravi90 replied:
Oh, you are correct, I do free it under the multi-threaded section, forgot to do it for the single threaded section, made the change now, thanks.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -996,84 +997,109 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+			tdata->policy = policy;
+			tdata->sidx = j;
+			j += chunk_size;
+			tdata->eidx = j - 1;
```
Now that you are filling out tdata twice in this function, I think it makes sense to make a function (even if the function has a thousand arguments). (same as in query_nodes())

agrawalravi90 commented on src/include/grunt.h:
```diff
@@ -62,6 +62,7 @@ extern int parse_node_resc(char *str, char **nodep, int *nl, struct key_value_pa
 extern int parse_node_resc_r(char *str, char **nodep, int *pnelem, int *nlkv, struct key_value_pair **kv);
 extern char *parse_plus_spec(char *selstr, int *rc);
 extern char *parse_plus_spec_r(char *selstr, char **last, int *hp);
+extern char *parse_plus_spec_mt_safe(char *selstr, char **tailptr, int *rc);
```
I thought _r() was for reentrant functions, which are often thread-safe but not necessarily (https://stackoverflow.com/questions/856823/threadsafe-vs-re-entrant). We already have a parse_plus_spec_r(), so I can't name it _r(). If you don't like _mt_safe(), could you please suggest some other name?

agrawalravi90 commented on src/lib/Libifl/grunt_parse.c:
```diff
@@ -650,3 +652,82 @@ parse_plus_spec(char *selstr, int *rc)
+ *	is grown as need to hold "selstr".
+ */
+char *
+parse_plus_spec_mt_safe(char *selstr, char **tailptr, int *rc)
```
hmm .. so parse_plus_spec_mt_safe() isn't similar parse_plus_spec_r(), it's actually similar to parse_plus_spec() instead. I see both parse_plus_spec() and parse_plus_spec_mt_safe() as convenience wrapper functions around parse_plus_spec_r() as they both call parse_plus_spec_r(). So, if I change parse_plus_spec_r() to also do the job that parse_plus_spec_mt_safe() does, then we'll have to create a new function which does what parse_plus_spec_r() currently does so that parse_plus_spec() can call that new function instead of parse_plus_spec_r(). Maybe it makes more sense to rename pbs_plus_spec_r() to  pbs_plus_spec_2() ? I think this is just a naming issue. Let me know what you think.

bhroam replied:
Like I said above, I think you should make one function that is MT safe.

agrawalravi90 replied:
But that's what I have right now. I needed a MT safe version of parse_plus_spec(), which is  what parse_plus_spec_mt_safe() is. parse_plus_spec_mt_safe() and parse_plus_spec() both call parse_plus_spec_r(), so parse_plus_spec_r() is a lower level function, that's why I suggested renaming it to parse_plus_spec_2().

bhroam replied:
No, you are creating a second function that does exactly what parse_plus_spec() does, but is MT safe.  What I am suggesting is making parse_plus_spec() MT safe, so we only have one version of the function.

agrawalravi90 replied:
Oh ok. But won't that make things slower for other daemons that call parse_plus_spec()? It will cause memory to be allocated on the heap every time instead of just once (https://stackoverflow.com/questions/2264969/why-is-memory-allocation-on-heap-much-slower-than-on-stack)

bhroam replied:
I'm not talking about allocating memory on the heap.  I'm talking abut turning parse_plus_spec() into something like strtok_r().  You pass an opaque pointer into the call for the function to use.

agrawalravi90 commented on src/lib/Libifl/grunt_parse.c:
```diff
@@ -650,3 +652,82 @@ parse_plus_spec(char *selstr, int *rc)
+ *
+ * @param[in] selstr - string holding select specs
+ * @param[in,out] tailptr - pointer to the end of the last substring
+ * @param[in] rc - flag
```
I copied the signature and docstring of parse_plus_spec(). I've changed it to an out, I'm not sure if I should change the name of the variable though, do we want to change it for parse_plus_spec() as well?

agrawalravi90 commented on src/scheduler/job_info.c:
```diff
@@ -996,84 +997,109 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+			tdata->oarr = NULL; /* Will be filled by the thread routine */
+			tdata->sinfo = qinfo->server;
+			tdata->qinfo = qinfo;
+			tdata->pbs_sd = pbs_sd;
```
We don't parallelize the statjob calls, we do a statjob before we parallelize the parsing of batch_status-es and creation of sched's job list(s)

bhroam replied:
Then why do we need to pass in pbs_sd?

agrawalravi90 replied:
Ah ok, I see your concern. Yes, the IFL calls become thread-safe after removing the call to pbs_client_thread_set_single_threaded_mode() from pbs_sched.c
pbs_sd is used for the following:
- Updating the job attributes estimated.soft_walltime, comment and accrue_type (pbs_alterjob calls)
- Setting 'peer_sd' on the job object if peer queue-ing

Please go through the code inside query_jobs_chunk() for more details, I just copy pasted most of the stuff from what was there before in query_jobs().

agrawalravi90 commented:
@bhroam Thanks for the review, I've addressed your comments, please provide further feedback, thanks!

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -6265,6 +6321,42 @@ check_node_eligibility_chunk(th_data_nd_eligible *data)
+	tdata->err = NULL;
+	tdata->pl = pl;
+	tdata->resresv = resresv;
+	tdata->exclerr_buf = exclerr_buf;
```
I thought I made a comment to not do pass this in.  Just translate it if you need it.

agrawalravi90 replied:
I must have missed it, will make the change thanks.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -6315,41 +6407,22 @@ check_node_array_eligibility(node_info **ninfo_arr, resource_resv *resresv, plac
 	} else {	 /* We are multithreading */
-		j = 0;
 		chunk_size = num_nodes / num_threads;
 		chunk_size = (chunk_size > 1024) ? chunk_size : 1024;
```
Make chunk size limits (1024 and 8192) macros.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -860,7 +860,7 @@ query_jobs_chunk(th_data_query_jinfo *data)
  * @retval a newly allocated th_data_query_jinfo object
  * @retval NULL for malloc error
  */
-static th_data_query_jinfo *
```
Make sure the inline keyword works for all the platforms PBS supports.  It came in the C99 standard.  It works with gcc, but will it work with Windows?

agrawalravi90 replied:
We are also removing Windows support for the scheduler with this PR, so Windows shouldn't be an issue. I'll do a build with all other platforms to be sure, thanks.

agrawalravi90 commented:
Thanks @bhroam I've addressed your comments, please provide further feedback.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -6307,6 +6304,9 @@ check_node_eligibility_chunk(th_data_nd_eligible *data)
 							n->nscr.ineligible = 1;
+							set_schd_error_codes(misc_err, NOT_RUN, NODE_NOT_EXCL);
+							translate_fail_code(misc_err, NULL, exclerr_buf);
+							clear_schd_error(misc_err);
```
Just make a call to schdlogerr() instead of doing the translate yourself.  It will not translate if we're not going to print it.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -6262,8 +6260,8 @@ check_node_eligibility_chunk(th_data_nd_eligible *data)
 	place *pl;
-	char *exclerr_buf;
 	node_info **ninfo_arr;
+	char exclerr_buf[MAX_LOG_SIZE] = {0};
```
No need to zero this.  translate_fail_code() does not require that.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -4932,9 +4933,16 @@ parse_execvnode(char *execvnode, server_info *sinfo)
 	}
 
-	simplespec = parse_plus_spec_mt_safe(execvnode, &tailptr, &ret);
+	excvndup = strdup(execvnode);
```
Use string_dup()

agrawalravi90 replied:
I've changed it. But I'm curious, why did we decide to use malloc + strcpy inside string_dup() instead of a strdup() ?

bhroam replied:
At the time I wrote the function, strdup() was not POSIX compliant.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -4933,12 +4933,7 @@ parse_execvnode(char *execvnode, server_info *sinfo)
-		log_err(errno, __func__, MEM_ERR_MSG);
-		return NULL;
-	}
+	excvndup = string_dup(execvnode);
```
string_dup() can still return NULL on a failed malloc().  It just handles the logging for you.

agrawalravi90 commented:
@bhroam thanks I've addressed your comments. Please let me know if it looks okay now.
@arungrover please review the new commits. I'd like to rebase this soon if possible because the number of conflicts are growing. Thanks!

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -4971,22 +4979,20 @@ parse_execvnode(char *execvnode, server_info *sinfo)
-				}
+			simplespec = parse_plus_spec_r(tailptr, &tailptr, &hp);
+			if (simplespec != NULL) {
+				int ret;
```
Do you need to free kv and excvndup here?

agrawalravi90 replied:
I don't need to free it inside the loop, I do it immediately outside of it.

mike0042 commented:
@agrawalravi90  - Is this PR still WIP with two approvals? If not, please update the title. Have comments from @subhasisb been addressed?

agrawalravi90 commented:
> @agrawalravi90 - Is this PR still WIP with two approvals? If not, please update the title. Have comments from @subhasisb been addressed?

Hey Mike, this PR is indeed WIP, I'm doing final testing and adding minor fixes as I go along.  It's taking time since this branch was not up to date with master for more than a month, and I need to test the MT performance on bigger machines. I was only able to do rebase after the 2 sign-offs. I'll remove WIP once I'm satisfied with it, thanks for checking!

agrawalravi90 commented:
@bhroam  and @arungrover  Please review the latest commit, I'm still doing some mem-leak testing, once I'm happy with it I'll remove the WIP tag and this will be ready for merge. Thanks for being patient with this PR.

bhroam commented:
I don't understand the latest commit.  It removes all the code which was put in to handle quickly deleting large numbers of jobs.  Now it just does the qdel.  Before it would kill the processes to speed the process along.

agrawalravi90 commented:
I wrote a self test to compare the performance and the difference was negligible:
[ravi@pbspro selftest]$ grep "Time taken for job cleanup" test_cleanup_after.log test_cleanup_before.log 
test_cleanup_after.log:2019-11-15 19:16:10,967 INFO     Time taken for job cleanup 817.227475643158
test_cleanup_before.log:2019-11-15 18:32:30,302 INFO     Time taken for job cleanup 798.5897672176361

The original approach was causing weird issues, the jobs were sometimes not getting deleted properly. Since the performance difference is negligible, I thought that the simpler, more reliable approach would be better

agrawalravi90 commented:
@bhroam  and @arungrover  I'm now happy with this PR, please review the latest commit(s) and let me know if I can rebase again since some conflicts seem to have popped up.

agrawalravi90 commented:
@PBSPro/pbspro-maintainers Finally, this PR is ready to be merged. @subhasisb your review is currently blocking this PR, do you have any further feedback for this PR?

agrawalravi90 commented:
@arungrover and @bhroam I reverted the change to cleanup_jobs() since it was going to slow down mom daemon restart, please review it once and, if it looks okay, then this is ready to be merged.

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8464,42 +8464,71 @@ def cleanup_jobs(self):
+                a = {'scheduling': 'False'}
+                self.manager(MGR_CMD_SET, SCHED, a, id=sc['id'],
+                             runas=ROOT_USER)
+                self.expect(SCHED, a, id=sc['id'])
```
I'm not sure what this accomplishes.  You just set it to false, won't manager throw an exception if it fails?

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8464,42 +8464,71 @@ def cleanup_jobs(self):
+                                    runas=ROOT_USER, logerr=False)
+            if running_job is True:
+                _msg = job_ids[-1] + ';'
+                _msg += 'Job Obit notice received has error 15001'
```
If you want to, you can fix this.  We have a ticket to fix it next sprint though.  Basically this will assume the last pid is running.  If it isn't, it will wait for a message that will never come for 100s.  To fix it, instead of doing job_ids[-1], you do a filter on state=R and take the last index of that list.  It's not hard, but we can fix it later.

agrawalravi90 replied:
It does seem straight-forward, but I'd prefer that we do it separately since this PR has been open for a pretty long time now.
